contests/_35-PDP/a-coupon-solution.md

```diff
@@ -0,0 +1,101 @@
+---
+layout: solution
+codename: coupon
+---
+
+## Επεξήγηση εκφώνησης
+Η συγκεκριμένη εκφώνηση είναι αρκετά ιδιαίτερη. Αν δεν την καταλάβετε στην παρακάτω μορφή, προτείνουμε να διαβάσετε την επίσημη εκφώνηση.
+
+Εν συντομία, μας δίνεται μια λίστα με $$N$$ αριθμούς $$C_1, \ldots C_N$$, ένας δεκαδικός αριθμός $$A$$ μεταξύ $$0$$ και $$1$$, κι ένας αριθμός $$B$$. Για κάθε ακέραιο $$X$$ ορίζουμε την παρακάτω διαδικασία (σε ψευδοκώδικα):
+
+```c++
+integer TotalMoney(integer X) {
+  integer s=0;
+  for(integer i=1; i<=N; ++i) {
+    if (X<10) X=0;
+    s = s + X*C[i];
+    X = A*X;    
+  }
+  return s;
+}
+```
+
+Ο στόχος είναι να βρούμε το μέγιστο $$X$$ έτσι ώστε $$TotalMoney(X) \le B$$.
+
+## Λύση 1Α - Γραμμική αναζήτηση από κάτω
+
+Είναι εύκολο να δούμε ότι όταν αυξάνουμε το $$X$$, αυξάνεται και η $$TotalMoney(X)$$ (για να είμαστε κάπως πιο ακριβείς: *δεν μειώνεται* η $$TotalMoney(X)$$).
+
+Μία πολύ απλή λύση είναι να δοκιμάσουμε όλα τα πιθανά $$X$$. Σταματάμε την πρώτη φορά που θα συναντήσουμε ένα $$X$$ για το οποίο $$TotalMoney(X)>B$$, και επιστρέφουμε το ακριβώς προηγούμενο $$X$$. Ο λόγος που σταματάμε και δεν ψάχνουμε για άλλα $$X$$ είναι ακριβώς το γεγονός ότι η $$TotalMoney(X)$$ δεν πρόκειται να μικρύνει όσο αυξάνουμε το $$X$$. Άρα για κάθε μεγαλύτερο $$X$$, η $$TotalMoney(X)$$ θα ήταν μεγαλύτερη από $$B$$, κι άρα δεν μας απασχολεί.
+
+Ενδεικτικά δίνουμε τον παρακάτω κώδικα για το πρόβλημα:
+{% include code.md solution_name='coupon_brute.cc' %}
+
+Μερικές επεξηγήσεις για τον κώδικα:
```

Comment:
Δεν ξέρω αν είναι καλύτερο αν κάποιες από αυτές τις επεξηγήσεις ήταν στον κώδικα

Reply:
Έβαλα επεξηγήσεις για long long και την for loop μέσα στον κώδικα, και έσβησα την επεξήγηση περί των print (περιττή). Στην επεξήγηση κράτησα μόνο τα περί long long ώστε να τονίσω ότι πρέπει να το σκέφτονται πάντα όταν υπάρχει πολλαπλασιασμός.